### PR TITLE
Expose remaining deep worksheet features over the v1 API

### PR DESCRIPTION
**Pull request by ggyuchive**
## Summary
Each feature is a validator plus a controller under a tab, mirroring the existing cell and worksheet-settings endpoints; GET returns plain JSON (a Yorkie value's toJSON would otherwise double-encode) and PUT/POST validate before writing.
- conditional-formats / data-validations (reuse the sheets normalizers)
- filter / pivot (null clears)
- range-styles / sheet-style
- column/row styles and column-widths / row-heights (1-based index keys, per-index merge, null clears an index)
- charts (PUT replaces the whole collection)
- clear-range (POST empties a range, keeping row/column structure)

## Why
Add REST /api/v1 endpoints for the spreadsheet features that were reachable in the app but not the public API, so external callers can control formatting and structure, not just cell values.

## Linked Issues

Fixes #

## Author checklist

- [X] I searched existing issues and PRs and confirmed this is not a duplicate.
- [ ] I read every line of this diff and can explain why each change is necessary.
- [X] Changes follow the conventions in the touched packages; any deviations are explained in *Why* above.
- [X] If AI tools assisted with this PR, I noted where in *Notes for Reviewers* below.

## Verification

Pass/fail is in this PR's own checks list. The comment CI posts reports
**scope** instead: which heavy jobs were skipped, and which changed file
forced the whole suite.

- [ ] verify-self — green in the checks list
- [ ] verify-integration — green in the checks list (or explicit skip reason below)

Skip reason (if applicable):

## Risk Assessment

- User-facing risk:
- Data/security risk:
- Rollback plan:

## Notes for Reviewers
Worked with Claude Opus 5
- UI changes (screenshots/gifs if applicable):
- Follow-up work (if any):

## Summary by CodeRabbit

* **New Features**
  * Added APIs for managing worksheet charts, filters, pivot tables, conditional formatting, data validation, styles, dimensions, and structure.
  * Added support for retrieving, replacing, merging, and clearing worksheet properties.
  * Added range-clearing functionality with the number of affected cells reported.
* **Bug Fixes**
  * Improved validation and error handling for invalid data, missing tabs, unsupported document types, and oversized ranges.
* **Tests**
  * Added comprehensive coverage for worksheet operations, persistence, serialization, and edge cases.

### CODE REVIEW

File: packages/backend/src/api/v1/worksheet-charts.controller.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    const out: SheetChart[] = [];
+    if (source && typeof source === 'object') {
+      for (const value of Object.values(source as object)) {
+        if (value && typeof value === 'object') {
+          out.push({ ...(value as SheetChart) });
+        }
+      }
+    }
+    return out;
+  }
```
[the remark this answers is not in the document]
Comment (hackerwins) [reply packages/backend/src/api/v1/worksheet-charts.controller.ts:76]: Confirmed, and fixed in b6ce5ec9b — but not the way this suggests, because `detachYorkieValue` does not fix it.

That helper branches on `Array.isArray(value)`, which is **false** for a Yorkie array proxy (its target is a `CRDTArray`, not an `Array`). It falls through to `Object.entries()` and emits `{createdAt, movedAt}` CRDT metadata where the array belonged. Applying it here would have swapped a crash for silently-wrong data.

The symptom is also worse than described. Reproduced against `@yorkie-js/sdk@0.7.17`: `CRDTArray.toJSON()` iterates the proxy, gets plain JS elements, and calls `.toJSON()` on them, so `GET /charts` throws `TypeError: value.toJSON is not a function` inside `res.json()` — an HTTP 500, not a stringified array.

The right helper was `unwrapJson`, in the same file. The object proxy's own `toJSON` walks `CRDTObject`'s element map rather than the proxy and returns correct nested JSON; `slides-tree.ts:53-68` already uses it for exactly this and says so.

Worth noting the sibling this comment points at as precedent — `worksheet-filter-pivot.controller.ts` — had the same bug, so `filter.hiddenRows` and all four pivot field arrays were already being served as CRDT metadata. Fixed in the same commit.

New regression tests seed a real `yorkie.Document` via `doc.update` (no server needed); the existing specs use `createSpreadsheetDocument()`, a plain object that can never be a proxy, so nothing here could have caught it.

File: packages/backend/src/api/v1/worksheet-structure.controller.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        let cleared = 0;
+        doc.update((root) => {
+          const ws = this.worksheetOrThrow(root, tabId);
+          for (const ref of toRefsFromRanges([range])) {
+            if (getWorksheetCell(ws, ref) !== undefined) {
+              writeWorksheetCell(ws, ref, undefined);
+              cleared++;
+            }
+          }
+        });
```
[the remark this answers is not in the document]
Comment (hackerwins) [reply packages/backend/src/api/v1/worksheet-structure.controller.ts:81]: Confirmed and fixed in 73f542947. Measured the controller's real per-cell work (`toRefsFromRanges` + `getWorksheetCell` against a populated worksheet) at ~1e6 cells per 0.5s, so a 1,000,000-cell cap is the limit; the full grid would otherwise be ~2.5 hours of blocked event loop from a ~20-byte body.

Followed the `parseMerges` precedent in `worksheet-settings.ts`, which already makes this argument and enforces `MaxRows` / `MaxColumns` / `MaxMergedCells`. The range is normalized **before** the check — `"C3:A1"` parses reversed and `toRefs` would still expand the full area otherwise.

One correction: `A1:XFD1048576` is Excel's grid, not wafflebase's (`MaxRows` is 1,000,000), so it is rejected by the bounds check and does not exercise the area cap. The over-area test uses an in-grid `A1:ZZ1000000` instead.

File: packages/backend/src/yorkie/worksheet-filter-pivot.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  if (filter === null || filter === undefined) return null;
+  const f = asObject(filter, 'filter must be an object or null');
+  const columns = asObject(f.columns ?? {}, 'filter.columns must be an object');
+  return {
+    startRow: nonNegInt(f.startRow, 'filter.startRow'),
+    endRow: nonNegInt(f.endRow, 'filter.endRow'),
+    startCol: nonNegInt(f.startCol, 'filter.startCol'),
+    endCol: nonNegInt(f.endCol, 'filter.endCol'),
+    columns,
+  } as WorksheetFilterState;
```
[the remark this answers is not in the document]
Comment (hackerwins) [reply packages/backend/src/yorkie/worksheet-filter-pivot.ts]: Confirmed and fixed in 58718989b — this was the most consequential finding on the PR.

`hiddenRows` is stored, not derived: `loadFilterState` (`sheet.ts:1536-1539`) trusts the persisted array and never recomputes it from the conditions. So a filter set through this endpoint opened in the editor with its dropdowns armed and its conditions visible, and **every row still showing**. No crash, because both readers defend with `|| []` — which is exactly why it would have shipped unnoticed.

Also dropped the blanket `as WorksheetFilterState`, keeping one narrow cast on `columns` so the documented "structural check only" contract for conditions survives. A future dropped field is now a compile error. `parsePivot` had the same cast and got the same treatment.

The existing test was a `toMatchObject` over four bounds, so it could not see a missing fifth field; it is a `toEqual` over the whole object now.

File: packages/backend/src/yorkie/worksheet-styles.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * Validate a `{ style: CellStyle | null }` body for the whole-sheet style.
+ * `null` clears it; an object is validated with `parseCellStyle`.
+ */
+export function parseSheetStyle(body: unknown): CellStyle | null {
+  if (typeof body !== 'object' || body === null || Array.isArray(body)) {
+    throw new BadRequestException('body must be an object { style: {...} | null }');
+  }
+  const style = (body as Record<string, unknown>).style;
+  if (style === null || style === undefined) return null;
+  return parseCellStyle(style);
```
[the remark this answers is not in the document]
Comment (hackerwins) [reply packages/backend/src/yorkie/worksheet-styles.ts:60]: Confirmed and fixed in 58718989b. Three things ruled out "this is the intended contract": the doc comment says only that `null` clears, the write is a merge rather than a replace (so "PUT replaces the resource" does not apply), and every sibling parser in this PR — `worksheet-dimensions.ts`, the `rangeStyles` path in this same file, `worksheet-rules.ts`, `worksheet-charts.ts` — already 400s on an omitted key.

`parseFilter` and `parsePivot` had the identical gap and are fixed in the same commit. Added a controller-level test asserting the stored style *survives* a rejected request — a 400 that still deleted it would pass a parser-only test.

File: packages/backend/src/yorkie/worksheet-filter-pivot.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (hackerwins) [reply packages/backend/src/yorkie/worksheet-filter-pivot.ts]: Fixed in 58718989b, though the stated impact is not right.

The validation gap is real and the ordering check is now there. But "the stored filter then describes an empty region" does not hold: the only reader hands the pair to `toRange()` (`coordinates.ts:97-108`), which normalizes with `Math.min`/`Math.max`, so an inverted rectangle loads as a well-formed one. The observable effect is just that `GET filter` echoes the inverted values back.

Taken because the doc comment claims the bounds are validated and they were not — contract honesty, not a crash.

File: packages/backend/src/yorkie/worksheet-dimensions.spec.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        parseIndexKeyedStyles({ columnStyles: [] }, 'columnStyles'),
+      ).toThrow(BadRequestException);
+    });
+
+    it('rejects a non-integer / non-positive index key', () => {
+      expect(() =>
+        parseIndexKeyedStyles({ columnStyles: { '0': { b: true } } }, 'columnStyles'),
+      ).toThrow(BadRequestException);
+      expect(() =>
+        parseIndexKeyedStyles({ columnStyles: { A: { b: true } } }, 'columnStyles'),
```
[the remark this answers is not in the document]
Comment (hackerwins) [reply packages/backend/src/yorkie/worksheet-dimensions.spec.ts]: Fixed in cf323b697, and the list needed correcting. Running `prettier --check` (3.5.3, the version this package pins) against all 25 changed `.ts` files: **16 fail**, not 9. The two flagged here that actually *pass* are `worksheet-styles.controller.ts` and `worksheet-dimensions.controller.ts`; nine failing files were not named.

All 16 are formatted now, and all 25 pass `prettier --check` and `eslint` clean.

Root cause for CI being green throughout: no lane runs `pnpm backend lint`. `verify-self` runs `lint:arch`, a separate config without the `prettier/prettier` rule. `main` carries ~50 files in the same state, so adding that gate means clearing those first — noted in the task doc, not attempted here.

File: packages/backend/src/api/v1/worksheet-filter-pivot.controller.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    return this.yorkieService.withDocument(
+      documentId,
+      (doc) => {
+        const ws = doc.getRoot().sheets?.[tabId] as
+          | { filter?: unknown }
+          | undefined;
+        return { filter: ws?.filter ? detachYorkieValue(ws.filter) : null };
+      },
+      { syncMode: 'readonly' },
+    );
```
[the remark this answers is not in the document]
Comment (hackerwins) [reply packages/backend/src/api/v1/worksheet-filter-pivot.controller.ts:78]: Not taken, deliberately — recorded in `docs/tasks/active/20260827-v1-worksheet-review-fix-todo.md`.

The behavior is as described, but the three sites anchored here are 3 of **15** identical new GET handlers (`worksheet-rules` ×2, `worksheet-styles` ×2, `worksheet-dimensions` ×8). More to the point, the repo is already split on this and both precedents predate the PR: `cells.controller.ts:60` throws `NotFoundException`, while `worksheet.controller.ts` — the controller these six were copied from, `worksheetOrThrow`-on-write-only included — does not.

Patching the three cited sites would leave the PR self-inconsistent. Making GET/PUT parity uniform across all 15 is worth doing, but as one pass with a decision about whether it extends to the three pre-existing handlers, since that changes public API behavior for current callers.

**Comment by hackerwins** [comment]
Review round addressed — 4 commits on `cf323b697`, rebased onto current `main` (which now carries #975/#976).

| Commit | What |
| --- | --- |
| `b6ce5ec9b` | `GET /charts` 500 on nested `seriesColumns`, plus the same bug already live in the filter/pivot handlers |
| `73f542947` | Area + grid bound on `clear-range` |
| `58718989b` | `hiddenRows` loss, inverted ranges, omitted-key-clears across `parseSheetStyle` / `parseFilter` / `parsePivot` |
| `cf323b697` | Prettier across all 25 files, 3 dead type assertions, task docs |

Each finding was written test-first against the unfixed code and shown to fail before it passed. `pnpm verify:fast` green on every commit; backend is 1134 tests / 88 suites, and all 25 changed `.ts` files pass `prettier --check` and `eslint`.

**One defect nobody raised.** The `detachYorkieValue` suggestion on `readCharts` pointed at the filter/pivot controller as the precedent to copy. That precedent had the same bug — `Array.isArray` is false for a Yorkie array proxy, so `filter.hiddenRows` and all four pivot field arrays were being served as `{createdAt, movedAt}` CRDT metadata. Silently wrong data, no error. `unwrapJson` is the correct helper (`slides-tree.ts` sets the precedent); details in the thread.

**Not taken**, both recorded in `docs/tasks/active/20260827-v1-worksheet-review-fix-todo.md` so they are not re-opened:

- *GET handlers skip the tab-existence check* — the 3 cited sites are 3 of 15 identical handlers, and the repo is already split (`cells.controller.ts` 404s, the pre-existing `worksheet.controller.ts` these were copied from does not). Fixing 3 makes the PR self-inconsistent; needs one uniform pass.
- *Duplicated `assertSheetDocument` / `worksheetOrThrow`* — real, and larger than reported: 8 files, ~150 lines, 2 of them predating this PR. Not verbatim either (message noun and return cast both vary), so extraction needs a parameter and a generic. Belongs in its own cleanup commit rather than widening this one.

Also worth flagging for a follow-up, out of scope here: `cells.controller.ts` `expandRange` has the same unbounded expansion on `GET ?range=` and is worse — it materializes a `string[]`, so it OOMs rather than stalls. Pre-existing, not introduced by this PR.
